File: src/agent_checkpoint/manifest.py
from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List, Optional, Union


@dataclass
class ConfidenceRange:
    """Range of confidence scores the agent operates within."""

    min: float = 0.0
    max: float = 1.0

    def __post_init__(self) -> None:
        if not (0.0 <= self.min <= 1.0):
            raise ValueError(f"min must be between 0.0 and 1.0; got {self.min}")
        if not (0.0 <= self.max <= 1.0):
            raise ValueError(f"max must be between 0.0 and 1.0; got {self.max}")
        if self.min > self.max:
            raise ValueError(f"min ({self.min}) cannot be greater than max ({self.max})")


@dataclass
class AgentManifest:
    """Manifest describing an agent's capabilities and constraints."""

    ac_version: str
    agent_id: str
    capabilities: List[str]
    refuses: List[str] = field(default_factory=list)
    confidence_range: Optional[ConfidenceRange] = None
    max_context_tokens: Optional[int] = None
    contact: Optional[str] = None


def parse_manifest(data: Union[str, Dict[str, Any]]) -> Optional[AgentManifest]:
    """
    Parses a JSON string or dictionary into an AgentManifest object.
    Returns None on any error or if required fields are missing.
    """
    try:
        if isinstance(data, str):
            raw_dict = json.loads(data)
        else:
            raw_dict = data

        if not isinstance(raw_dict, dict):
            return None

        # Check required fields
        if "ac_version" not in raw_dict or "agent_id" not in raw_dict or "capabilities" not in raw_dict:
            return None

        conf_range = None
        raw_conf = raw_dict.get("confidence_range")
        if isinstance(raw_conf, dict):
            try:
                conf_range = ConfidenceRange(
                    min=float(raw_conf.get("min", 0.0)), max=float(raw_conf.get("max", 1.0))
                )
            except (ValueError, TypeError):
                # If confidence range is invalid, we continue without it
                pass

        return AgentManifest(
            ac_version=str(raw_dict["ac_version"]),
            agent_id=str(raw_dict["agent_id"]),
            capabilities=list(raw_dict["capabilities"]),
            refuses=list(raw_dict.get("refuses", [])),
            confidence_range=conf_range,
            max_context_tokens=raw_dict.get("max_context_tokens"),
            contact=raw_dict.get("contact"),
        )
    except Exception:
        # Never raise, return None on any failure
        return None


def emit_manifest(manifest: AgentManifest) -> str:
    """
    Converts an AgentManifest to a JSON string.
    Omits fields with None values.
    """
    raw = asdict(manifest)

    def _remove_none(obj: Any) -> Any:
        if isinstance(obj, dict):
            return {k: _remove_none(v) for k, v in obj.items() if v is not None}
        return obj

    clean = _remove_none(raw)
    return json.dumps(clean, indent=2)

File: packages/ts/src/manifest.ts
import { AgentManifest, ConfidenceRange } from "./types.js";

export function parseManifest(data: string | object): AgentManifest | null {
  try {
    let obj: Record<string, unknown>;
    if (typeof data === "string") {
      obj = JSON.parse(data);
    } else {
      obj = data as Record<string, unknown>;
    }

    if (!obj || typeof obj !== "object" || Array.isArray(obj)) return null;

    if (typeof obj.ac_version !== "string" || typeof obj.agent_id !== "string" || !Array.isArray(obj.capabilities)) {
      return null;
    }

    let confidenceRange: ConfidenceRange | undefined;
    if (obj.confidence_range && typeof obj.confidence_range === "object" && !Array.isArray(obj.confidence_range)) {
      const cr = obj.confidence_range as Record<string, unknown>;
      const min = Number(cr.min ?? 0.0);
      const max = Number(cr.max ?? 1.0);

      if (!isNaN(min) && !isNaN(max) && min >= 0 && min <= 1 && max >= 0 && max <= 1 && min <= max) {
        confidenceRange = { min, max };
      }
    }

    return {
      ac_version: obj.ac_version,
      agent_id: obj.agent_id,
      capabilities: obj.capabilities as string[],
      refuses: Array.isArray(obj.refuses) ? (obj.refuses as string[]) : [],
      confidence_range: confidenceRange,
      max_context_tokens: typeof obj.max_context_tokens === "number" ? obj.max_context_tokens : undefined,
      contact: typeof obj.contact === "string" ? obj.contact : undefined,
    };
  } catch {
    return null;
  }
}

export function emitManifest(manifest: AgentManifest): string {

  function removeNone(obj: unknown): unknown {
    if (obj !== null && typeof obj === "object" && !Array.isArray(obj)) {
      const result: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(obj)) {
        if (v !== undefined && v !== null) {
          result[k] = removeNone(v);
        }
      }
      return result;
    }
    return obj;
  }

  return JSON.stringify(removeNone(manifest), null, 2);
}
